README.md
# [IntegrityChain]: FIPS 203 Module-Lattice-Based Key-Encapsulation Mechanism Standard

[![crate][crate-image]][crate-link]
[![Docs][docs-image]][docs-link]
[![Build Status][build-image]][build-link]
![Apache2/MIT licensed][license-image]
![Rust Version][rustc-image]

[FIPS 203] (Initial Public Draft) Module-Lattice-Based Key-Encapsulation Mechanism
Standard written in pure Rust.

This library implements the FIPS 203 **draft** standard in pure Rust with minimal and
mainstream dependencies. All three security parameter sets are fully functional. The
code does not require the standard library, e.g. `#[no_std]`, and has no heap
allocations so will be suitable for WASM, embedded and bare-metal applications.
Significant performance optimizations are forthcoming.

See <https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf> for a full
description of the target functionality.

The functionality is extremely simple to use, as demonstrated by the following example.

~~~rust
// Use the desired target parameter set.
use ml_kem_rs::ml_kem_512; // Could also be ml_kem_768 or ml_kem_1024. 

// Alice runs `key_gen()` and then serializes the encaps key `ek` for Bob (to bytes).
let (alice_ek, alice_dk) = ml_kem_512::key_gen();
let alice_ek_bytes = alice_ek.to_bytes();

// Alice sends the encaps key `ek_bytes` to Bob.
let bob_ek_bytes = alice_ek_bytes;

// Bob deserializes the encaps `ek_bytes` and then runs `encaps() to get the shared 
// secret `ssk` and ciphertext `ct`. He serializes the ciphertext `ct` for Alice (to bytes).
let bob_ek = ml_kem_512::new_ek(bob_ek_bytes);
let (bob_ssk_bytes, bob_ct) = bob_ek.encaps();
let bob_ct_bytes = bob_ct.to_bytes();

// Bob sends the ciphertext `ct_bytes` to Alice
let alice_ct_bytes = bob_ct_bytes;

// Alice deserializes the ciphertext `ct` and runs `decaps()` with her decaps key
let alice_ct = ml_kem_512::new_ct(alice_ct_bytes);
let alice_ssk_bytes = alice_dk.decaps(&alice_ct);

// Alice and Bob will now have the same secret key
assert_eq!(bob_ssk_bytes, alice_ssk_bytes);
~~~

Rust [Documentation][docs-link]

## Security Notes

This crate is functional and corresponds to the first initial public draft of FIPS 203.
This crate is still under construction/refinement -- USE AT YOUR OWN RISK!

## Supported Parameter Sets

- ML-KEM-512
- ML-KEM-768
- ML-KEM-1023

## Minimum Supported Rust Version

Rust **1.72** or higher.

Minimum supported Rust version can be changed in the future, but it will be
done with a minor version bump.

## SemVer Policy

- All on-by-default features of this library are covered by SemVer
- MSRV is considered exempt from SemVer as noted above

## License

All crates licensed under either of

* [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)
* [MIT license](http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.

[//]: # (badges)

[crate-image]: https://buildstats.info/crate/ml-kem-rs

[crate-link]: https://crates.io/crates/ml-kem-rs

[docs-image]: https://docs.rs/ml-kem-rs/badge.svg

[docs-link]: https://docs.rs/ml-kem-rs/

[build-image]: https://github.com/integritychain/ml-kem-rs/workflows/test/badge.svg

[build-link]: https://github.com/integritychain/ml-kem-rs/actions?query=workflow%3Atest

[license-image]: https://img.shields.io/badge/license-Apache2.0/MIT-blue.svg

[rustc-image]: https://img.shields.io/badge/rustc-1.72+-blue.svg

[//]: # (general links)

[IntegrityChain]: https://github.com/integritychain/

[FIPS 203]: https://csrc.nist.gov/pubs/fips/203/ipd


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.1.0 (2023-10-15)

- Initial release


src/lib.rs
#![no_std]
#![deny(clippy::pedantic)]
#![deny(warnings)]
#![deny(missing_docs)]
#![doc = include_str!("../README.md")]

/// Implements FIPS 203 draft Module-Lattice-based Key-Encapsulation Mechanism Standard.
/// See <https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf>

#[cfg(test)]
extern crate alloc;

// Supports automatically clearing sensitive data on drop
use zeroize::{Zeroize, ZeroizeOnDrop};

// Functionality map per FIPS 203 draft
//
// Algorithm 2 BitsToBytes(b) on page 17                    --> byte_fns.rs
// Algorithm 3 BytesToBits(B) on page 18                    --> byte_fns.rs
// Algorithm 4 ByteEncoded(F) on page 19                    --> byte_fns.rs
// Algorithm 5 ByteDecoded(B) on page 19                    --> byte_fns.rs
// Algorithm 6 SampleNTT(B) on page 20                      --> sampling.rs
// Algorithm 7 SamplePolyCBDη(B) on page 20                 --> sampling.rs
// Algorithm 8 NTT(f) on page 22                            --> ntt.rs
// Algorithm 9 NTT−1(fˆ) on page 23                         --> ntt.rs
// Algorithm 10 MultiplyNTTs(fˆ,ĝ) on page 24               --> ntt.rs
// Algorithm 11 BaseCaseMultiply(a0,a1,b0,b1,γ) on page 24  --> ntt.rs
// Algorithm 12 K-PKE.KeyGen() on page 26                   --> k_pke.rs
// Algorithm 13 K-PKE.Encrypt(ekPKE,m,r) on page 27         --> k_pke.rs
// Algorithm 14 K-PKE.Decrypt(dkPKE,c) on page 28           --> k_pke.rs
// Algorithm 15 ML-KEM.KeyGen() on page 29                  --> ml_kem.rs
// Algorithm 16 ML-KEM.Encaps(ek) on page 30                --> ml_ke.rs
// Algorithm 17 ML-KEM.Decaps(c,dk) on page 32              --> ml_kem.rs
// PRF and XOF on page 16                                   --> helpers.rs
// Three has functions: G, H, J on page 17                  --> helpers.rs
// Compress and Decompress on page 18                       --> helpers.rs
//
// The three parameter sets are modules in this file with injected macro code
// that connects them into the functionality in ml_kem.rs

mod byte_fns;
mod helpers;
mod k_pke;
mod ml_kem;
mod ntt;
mod sampling;
mod types;

#[cfg(test)]
mod smoke_test;

// Relevant to all parameter sets
const _N: u32 = 256;
const Q: u32 = 3329;
const ZETA: u32 = 17;
const SSK_LEN: usize = 32;

// Relevant to all parameter sets
/// The (opaque) secret key that can be deserialized by each party.
#[derive(Debug, Zeroize, ZeroizeOnDrop)]
pub struct SharedSecretKey([u8; SSK_LEN]);

impl SharedSecretKey {
    #[must_use]
    /// The `to_bytes` function deserializes a shared secret key into a byte array.
    pub fn to_bytes(&self) -> [u8; SSK_LEN] { self.0 }
}

// Conservative (constant-time) paranoia...
impl PartialEq for SharedSecretKey {
    fn eq(&self, other: &Self) -> bool {
        let mut result = true;
        for i in 0..self.0.len() {
            result &= self.0[i] == other.0[i];
        }
        result
    }
}


// This common functionality is injected into each parameter set module
macro_rules! functionality {
    () => {
        // TODO: Implement a 'global scratch' struct rather than the weird lower-level stuff here
        const ETA1_64: usize = ETA1 * 64; // Currently, Rust does not allow expressions involving
        const ETA1_512: usize = ETA1 * 512; // constants in type expressions such as [u8, ETA1 * 64].
        const ETA2_64: usize = ETA2 * 64; // So this is handled manually...what a pain
        const ETA2_512: usize = ETA2 * 512;
        const DU_256: usize = DU * 256;
        const DV_256: usize = DV * 256;
        const J_LEN: usize = 32 + 32 * (DU * K + DV);

        use rand::random;
        use zeroize::{Zeroize, ZeroizeOnDrop};

        /// Correctly sized encapsulation key specific to the target parameter set.
        #[derive(Zeroize, ZeroizeOnDrop)]
        pub struct EncapsKey([u8; EK_LEN]);

        /// Correctly sized decapsulation key specific to the target parameter set.
        #[derive(Zeroize, ZeroizeOnDrop)]
        pub struct DecapsKey([u8; DK_LEN]);

        /// Correctly sized ciphertext specific to the target parameter set.
        #[derive(Zeroize, ZeroizeOnDrop)]
        pub struct CipherText([u8; CT_LEN]);

        /// Per FIPS 203, the key generation algorithm `ML-KEM.KeyGen` for ML-KEM (Algorithm 15)
        /// accepts no input, utilizes randomness, and produces an encapsulation key and a
        /// decapsulation key. While the encapsulation key can be made public, the decapsulation key
        /// must remain private. This outputs of this function are opaque structs specific to a
        /// target parameter set.
        #[must_use]
        pub fn key_gen() -> (EncapsKey, DecapsKey) {
            let (mut ek, mut dk) = (EncapsKey::default(), DecapsKey::default());
            let random_z = random::<[u8; 32]>();
            let random_d = random::<[u8; 32]>();
            ml_kem::ml_kem_key_gen::<K, ETA1, ETA1_64, ETA1_512>(
                &random_z, &random_d, &mut ek.0, &mut dk.0,
            );
            (ek, dk)
        }

        /// Test only access to key generation seed
        #[must_use]
        #[cfg(test)]
        pub fn key_gen_test(seed: &[u8; 32]) -> (EncapsKey, DecapsKey) {
            let (mut ek, mut dk) = (EncapsKey::default(), DecapsKey::default());
            ml_kem::ml_kem_key_gen::<K, ETA1, ETA1_64, ETA1_512>(
                &seed, &seed, &mut ek.0, &mut dk.0,
            );
            (ek, dk)
        }

        /// The `new_ek` function deserializes a byte array of the correct length into an
        /// encapsulation key. The correct length of the input byte array is specific to a target
        /// parameter set and the output is an opaque struct.
        #[must_use]
        pub fn new_ek(bytes: [u8; EK_LEN]) -> EncapsKey { EncapsKey(bytes) }

        /// The `new_ct` function deserializes a byte array of the correct length into an opaque
        /// cipher text value. The correct length of the input byte array is specific to a target
        /// parameter set and the output is an opaque struct.
        #[must_use]
        pub fn new_ct(bytes: [u8; CT_LEN]) -> CipherText { CipherText(bytes) }

        impl EncapsKey {
            fn default() -> Self { EncapsKey([0u8; EK_LEN]) }

            /// Per FIPS 203, the encapsulation algorithm `ML-KEM.Encaps` of ML-KEM (Algorithm 16)
            /// accepts an encapsulation key as input, requires randomness, and outputs a ciphertext
            /// and a shared key. The inputs and outputs to this function are opaque structs
            /// specific to a target parameter set.
            #[must_use]
            pub fn encaps(&self) -> (SharedSecretKey, CipherText) {
                let mut ct = CipherText::default();
                let random_m = random::<[u8; 32]>();
                let ssk = ml_kem::ml_kem_encaps::<
                    K,
                    ETA1,
                    ETA1_64,
                    ETA1_512,
                    ETA2,
                    ETA2_64,
                    ETA2_512,
                    DU,
                    DU_256,
                    DV,
                    DV_256,
                >(&random_m, &self.0, &mut ct.0);
                (ssk, ct)
            }

            /// Test only access to encapsulation seed
            #[must_use]
            pub fn encaps_test(&self, seed: &[u8; 32]) -> (SharedSecretKey, CipherText) {
                let mut ct = CipherText::default();
                let ssk = ml_kem::ml_kem_encaps::<
                    K,
                    ETA1,
                    ETA1_64,
                    ETA1_512,
                    ETA2,
                    ETA2_64,
                    ETA2_512,
                    DU,
                    DU_256,
                    DV,
                    DV_256,
                >(&seed, &self.0, &mut ct.0);
                (ssk, ct)
            }

            /// The `to_bytes` function deserializes an encapsulation key into a byte array.
            #[must_use]
            pub fn to_bytes(&self) -> [u8; EK_LEN] { self.0.clone() }
        }


        impl DecapsKey {
            fn default() -> Self { DecapsKey([0u8; DK_LEN]) }

            /// Per FIPS 203, the decapsulation algorithm ML-KEM.Decaps of ML-KEM (Algorithm 16)
            /// accepts a decapsulation key and a ML-KEM ciphertext as input, does not use any
            /// randomness, and outputs a shared secret. The inputs and outputs to this function are
            /// opaque structs specific to a target parameter set.
            #[must_use]
            pub fn decaps(&self, ct: &CipherText) -> SharedSecretKey {
                ml_kem::ml_kem_decaps::<
                    K,
                    ETA1,
                    ETA1_64,
                    ETA1_512,
                    ETA2,
                    ETA2_64,
                    ETA2_512,
                    DU,
                    DU_256,
                    DV,
                    DV_256,
                    J_LEN,
                    CT_LEN,
                >(&self.0, &ct.0)
            }

            /// The `to_bytes` function deserializes a cipher text into a byte array.
            #[must_use]
            #[cfg(test)]
            pub fn to_bytes_test(&self) -> [u8; DK_LEN] { self.0.clone() }
        }


        impl CipherText {
            fn default() -> Self { CipherText([0u8; CT_LEN]) }

            /// The `to_bytes` function deserializes a cipher text into a byte array.
            #[must_use]
            pub fn to_bytes(&self) -> [u8; CT_LEN] { self.0.clone() }
        }
    };
}


///  ML-KEM-512 is claimed to be in security category 1, see table 2 & 3 on page 33.
#[cfg(feature = "ml_kem_512")]
pub mod ml_kem_512 {
    //!
    //! See <https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf>
    //!
    //! Typical usage flow entails:
    //! 1. The originator runs `key_gen()` to get an encaps key `encapsKey` and decaps key `decapsKey`.
    //! 2. The originator deserializes the encaps key via `encapsKey.to_bytes()` and sends to the remote party.
    //! 3. The remote party serializes the bytes via `new_ek(<bytes>)` to get the shared secret key `ssk` and ciphertext `cipherText`.
    //! 4. The remote party deserializes the cipertext via `cipherText.to_bytes()` and sends to the originator.
    //! 5. The originator serializes the ciphertext via `new_ct(<bytes>)` then runs `decapsKey.decaps(cipherText)` to the get shared secret ket `ssk`.
    //! 6. Both the originator and remote party now have the same shared secret key `ssk`.

    use crate::{ml_kem, SharedSecretKey};

    const K: usize = 2;
    const ETA1: usize = 3;
    const ETA2: usize = 2;
    const DU: usize = 10;
    const DV: usize = 4;
    const EK_LEN: usize = 800;
    const DK_LEN: usize = 1632;
    const CT_LEN: usize = 768;

    functionality!();
}


/// ML-KEM-768 is claimed to be in security category 3, see table 2 & 3 on page 33.
#[cfg(feature = "ml_kem_768")]
pub mod ml_kem_768 {
    //!
    //! See <https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf>
    //!
    //! Typical usage flow entails:
    //! 1. The originator runs `key_gen()` to get an encaps key `encapsKey` and decaps key `decapsKey`.
    //! 2. The originator deserializes the encaps key via `encapsKey.to_bytes()` and sends to the remote party.
    //! 3. The remote party serializes the bytes via `new_ek(<bytes>)` to get the shared secret key `ssk` and ciphertext `cipherText`.
    //! 4. The remote party deserializes the cipertext via `cipherText.to_bytes()` and sends to the originator.
    //! 5. The originator serializes the ciphertext via `new_ct(<bytes>)` then runs `decapsKey.decaps(cipherText)` to the get shared secret ket `ssk`.
    //! 6. Both the originator and remote party now have the same shared secret key `ssk`.

    use crate::{ml_kem, SharedSecretKey};

    const K: usize = 3;
    const ETA1: usize = 2;
    const ETA2: usize = 2;
    const DU: usize = 10;
    const DV: usize = 4;
    const EK_LEN: usize = 1184;
    const DK_LEN: usize = 2400;
    const CT_LEN: usize = 1088;

    functionality!();
}


/// ML-KEM-1024 is claimed to be in security category 5, see table 2 & 3 on page 33.
#[cfg(feature = "ml_kem_1024")]
pub mod ml_kem_1024 {
    //!
    //! See <https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf>
    //!
    //! Typical usage flow entails:
    //! 1. The originator runs `key_gen()` to get an encaps key `encapsKey` and decaps key `decapsKey`.
    //! 2. The originator deserializes the encaps key via `encapsKey.to_bytes()` and sends to the remote party.
    //! 3. The remote party serializes the bytes via `new_ek(<bytes>)` to get the shared secret key `ssk` and ciphertext `cipherText`.
    //! 4. The remote party deserializes the cipertext via `cipherText.to_bytes()` and sends to the originator.
    //! 5. The originator serializes the ciphertext via `new_ct(<bytes>)` then runs `decapsKey.decaps(cipherText)` to the get shared secret ket `ssk`.
    //! 6. Both the originator and remote party now have the same shared secret key `ssk`.

    use crate::{ml_kem, SharedSecretKey};

    const K: usize = 4;
    const ETA1: usize = 2;
    const ETA2: usize = 2;
    const DU: usize = 11;
    const DV: usize = 5;
    const EK_LEN: usize = 1568;
    const DK_LEN: usize = 3168;
    const CT_LEN: usize = 1568;

    functionality!();
}
